### DISCUSSION TITLE
Point metadata

### DISCUSSION THREAD

**Discussion by ilyakorsunsky** [CGAL/cgal#6940] [Q&A] [answered]
Hi, 

I am a first time user of CGAL. It's already speeding up some important calculations, so I'm excited to keep using it! What I need to do is to add some metadata (fields that don't affect geometric computations) to my Simple_cartesian::Point_2 objects. Is there an easy way to do this? 

Thanks! 

Ilya

**Comment by MaelRL** [comment]
You can create a map from some kernel objects to your data, or do something like:
```cpp
#include <CGAL/Simple_cartesian.h>
#include <CGAL/IO/Color.h>

using K = CGAL::Simple_cartesian<double>;
using Point_2 = K::Point_2;

struct Point_with_color
{
  Point_with_color(const Point_2& p) : p(p), c(CGAL::IO::white()) { }
  Point_with_color(const Point_2& p, const CGAL::IO::Color& c) : p(p), c(c) { }

  // Conversion to Point_2
  operator const Point_2&() const { return p; }

  Point_2 p;
  CGAL::IO::Color c;
};

int main()
{
  K k;

  Point_2 bp0(0,0), bp1(1,0), bp2(1,1);
  Point_with_color p0(bp0, CGAL::IO::red());
  Point_with_color p1(bp1, CGAL::IO::green());
  Point_with_color p2(bp2, CGAL::IO::blue());

  std::cout << "Are these points collinear? " << k.collinear_2_object()(p0, p1, p2) << std::endl;

  Point_with_color cc = k.construct_circumcenter_2_object()(p0, p1, p2);
  std::cout << "Circumcenter at pos: " << cc.p << " and color " << cc.c << std::endl;

  return 0;
}
```

Also, if you are using some specific data structures, there might also be more natural ways to associate extra information to your elements (such as [Triangulation_vertex_base_with_info_2](https://doc.cgal.org/latest/Triangulation_2/classCGAL_1_1Triangulation__vertex__base__with__info__2.html)).

**Comment by ilyakorsunsky** [comment]
Thanks for the suggestion @MaelRL! 

I intend to use these modified Point objects in a kd tree and then perform a range search with a fuzzy_sphere object. The code below seems to be convert the Point_with_index object back into a Point_2 object during the search, removing the additional meta data from the point. Do you have a suggestion for how to define fuzzy_sphere in a way that keeps the Point_with_index object intact? 

```
#include <CGAL/Simple_cartesian.h>
#include <CGAL/Kd_tree.h>
#include <CGAL/algorithm.h>
#include <CGAL/Fuzzy_sphere.h>
#include <CGAL/Search_traits_2.h>
#include <Rcpp.h>
using namespace Rcpp;

using K = CGAL::Simple_cartesian<double>;
using Point_2 = K::Point_2;
using Traits = CGAL::Search_traits_2<K>;
using Fuzzy_circle = CGAL::Fuzzy_sphere<Traits>;
using Tree = CGAL::Kd_tree<Traits>;

// [[Rcpp::depends(RcppCGAL)]]
// [[Rcpp::depends(RcppEigen)]]

struct Point_with_index
{
  Point_with_index(const Point_2& p) : p(p), index(0) { }
  Point_with_index(const Point_2& p, const int& i) : p(p), index(i) { }
  // Point_with_index(const Point_2& p, const int& i, const int& t) : p(p), index(i), type(t) { }

  // Conversion to Point_2
  operator const Point_2&() const { return p; }

  Point_2 p;
  int index;
  // int type; 
};

// [[Rcpp::export]]
IntegerMatrix count_neighbors_multitype(NumericVector xymat, size_t n, float radius, float epsilon, IntegerVector types, size_t ntypes)
{
  // build tree 
  Tree tree;
  for(int i = 0; i < n; i++)
    tree.insert(Point_with_index(Point_2(xymat(i, 0), xymat(i, 1)), i));
  tree.build(); 

  // for each point, count number of points in radius 10 
  // size_t count;
  IntegerMatrix n_neighbors(n, ntypes); 

  for(unsigned i = 0; i < n; i++) {
    Fuzzy_circle search_circle(
      Point_with_index(Point_2(xymat(i, 0), xymat(i, 1))),
      radius, 
      epsilon
    ); // center, radius, fuzziness 

    std::list<Point_with_index> result;
    tree.search(std::back_inserter(result), search_circle); 

    for (std::list<Point_with_index>::iterator it = result.begin(); it != result.end(); ++it) {
      // std::cout << it->index << std::endl; 
      n_neighbors(i, types(it->index))++; 
    }
  }
  return n_neighbors;   
}
```

**Comment by ilyakorsunsky** [comment]
@MaelRL, I also tried out the solution with mapping from Kernel objects to my data. If I understand correctly, the idea is to map from each Point_2 object's memory address to my data. Here, I'm having trouble because the Point_2 object seems to get copied (rather than assigned by reference) during node insertion into the tree and into the output iterator during each search through the tree. Is this copying something that can be easily avoided? I really appreciate your help with this!

**Comment by sloriot** [comment, marked as answer]
You can use something like this example: https://doc.cgal.org/latest/Spatial_searching/Spatial_searching_2searching_with_point_with_info_inplace_8cpp-example.html

**Comment by ilyakorsunsky** [comment]
Perfect, thank you!
